src/software/new_geom/vector.h

```diff
@@ -0,0 +1,510 @@
+#pragma once
+
+#include <boost/geometry.hpp>
+#include <cmath>
+#include <iostream>
+
+#include "software/new_geom/angle.h"
+
+/**
+ * A vector in 2D space.
+ */
+class Vector final
+{
+   public:
+    /**
+     * Creates a zero-vector.
+     */
+    explicit constexpr Vector();
+
+    /**
+     * Creates a unit-magnitude Vector from an angle.
+     *
+     * @param angle the angle
+     *
+     * @return Vector the Vector
+     */
+    static Vector createFromAngle(const Angle &angle);
+
+    /**
+     * Creates a Vector with arbitrary x and y values.
+     *
+     * @param x the <var>x</var> value of the Vector
+     * @param y the <var>y</var> value of the Vector
+     */
+    constexpr Vector(double x, double y);
+
+    /**
+     * Creates a new Vector that is a copy of the given Vector
+     *
+     * @param the Vector to duplicate
+     */
+    constexpr Vector(const Vector &v);
+
+    /**
+     * Returns the magnitude in the x-coordinate of this Vector
+     *
+     * @return the magnitude in the x-coordinate of this Vector
+     */
+    constexpr double x() const;
+
+    /**
+     * Returns the magnitude in the y-coordinate of this Vector
+     *
+     * @return the magnitude in the y-coordinate of this Vector
+     */
+    constexpr double y() const;
+
+    /**
+     * Sets the magnitudes of this vector to the new magnitudes
+     *
+     * @param x the new magnitude in the x-coordinate
+     * @param y the new magnitude in the y-coordinate
+     */
+    void set(double x, double y);
+
+    /**
+     * Sets the magnitude in the x-coordinate of this vector
+     *
+     * @param x the new x-coordinate magnitude
+     */
+    void setX(double x);
+
+    /**
+     * Sets the magnitude in the y-coordinate of this vector
+     *
+     * @param y the new y-coordinate magnitude
+     */
+    void setY(double y);
+
+    /**
+     * Returns the square of the length of the Vector
+     *
+     * @return the square of the length of the Vector
+     */
+    constexpr double lensq() const;
+
+    /**
+     * Returns the length of the Vector
+     *
+     * @return the length of the Vector
+     */
+    double len() const;
+
+    /**
+     * Returns this Vector as a unit vector in the same direction
+     *
+     * @return Vector a unit vector in the same direction as this Vector, or a
+     * zero-vector if this Vector is a zero-vector
+     */
+    Vector norm() const;
+
+    /**
+     * Returns a scaled normalized vector in the same direction as this
+     * Vector
+     *
+     * @param length the desired length of the resultant vector
+     *
+     * @return a vector in the same direction as this Vector and with the given length,
+     * or a zero-vector if this Vector is a zero-vector
+     */
+    Vector norm(double length) const;
+
+    /**
+     * Returns the vector perpendicular to this Vector
```

Comment:
Do we want to formalize any convention for which perpendicular vector is returned here? I happen to know it's rotated positive 90 degrees from the original vector (so pointing perpendicular left), so should we formalize that in the comment or do people have other opinions on how perpendicular vectors should work?

@garethellis0 @jonathanlew @jsn0716

Reply:
Has this been resolved already? I think it should either be specified in the comment to be a positive 90 degrees rotation or take a bool parameter to rotate one way or the other.

Reply:
Has this been resolved? I think it should either be specified in the comment that it's a positive rotation or there should be a bool parameter to decide which direction

Reply:
Yes it's resolved. GitHub doesn't show the changes in the conversation comment, it just marks it as "outdated" at the top. If you look at the latest set of changes it should show the new comment.